From 4d6b6afee60189eeef7c89990a7c0120a07c4af5 Mon Sep 17 00:00:00 2001
From: Mark Yen <marky+git@activestate.com>
Date: Tue, 14 Feb 2012 16:21:14 -0800
Subject: [PATCH] Bug #3487440. Fix bug where setting an indicator on for whole document

---
 src/Decoration.cxx |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git src/Decoration.cxx src/Decoration.cxx
index 90bde57..8965883 100644
--- src/Decoration.cxx
+++ src/Decoration.cxx
@@ -28,7 +28,7 @@ Decoration::~Decoration() {
 }
 
 bool Decoration::Empty() {
-	return rs.Runs() == 1;
+	return (rs.Runs() == 1) && (rs.AllSameAs(0));
 }
 
 DecorationList::DecorationList() : currentIndicator(0), currentValue(1), current(0),
@@ -144,7 +144,7 @@ void DecorationList::DeleteRange(int position, int deleteLength) {
 void DecorationList::DeleteAnyEmpty() {
 	Decoration *deco = root;
 	while (deco) {
-		if (deco->Empty()) {
+		if ((lengthDocument == 0) || deco->Empty()) {
 			Delete(deco->indicator);
 			deco = root;
 		} else {
-- 
1.7.3.1.msysgit.0

