Index: komodo-devel/src/scintilla/macosx/ScintillaMacOSX.h
===================================================================
--- macosx/ScintillaMacOSX.h	(revision 44591)
+++ macosx/ScintillaMacOSX.h	(working copy)
@@ -110,6 +110,7 @@
 
     /// Register the notify callback
     void registerNotifyCallback(intptr_t windowid, SciNotifyFunc callback);
+    void unregisterNotifyCallback();
 private:
     virtual void Initialise();
     virtual void Finalise();
Index: komodo-devel/src/scintilla/macosx/ScintillaMacOSX.cxx
===================================================================
--- macosx/ScintillaMacOSX.cxx	(revision 44591)
+++ macosx/ScintillaMacOSX.cxx	(working copy)
@@ -1310,6 +1344,11 @@
     notifyProc = callback;
 }
 
+void ScintillaMacOSX::unregisterNotifyCallback() {
+    notifyObj = NULL;
+    notifyProc = NULL;
+}
+
 void ScintillaMacOSX::NotifyParent(SCNotification scn) {
     if (NULL != notifyProc) {
         scn.nmhdr.hwndFrom = (void*) this;
