Index: /home/toddw/as/komodo-devel/src/scintilla/win32/ScintillaWin.cxx
===================================================================
--- win32/ScintillaWin.cxx	(revision 42054)
+++ win32/ScintillaWin.cxx	(working copy)
@@ -1113,12 +1133,14 @@
 void ScintillaWin::SetMouseCapture(bool on) {
 	if (mouseDownCaptures) {
 		if (on) {
 			::SetCapture(MainHWND());
 		} else {
 			::ReleaseCapture();
+                        // KOMODO: Disable drag drop tracking - bug 87342.
+			inDragDrop = ddNone;
 		}
 	}
 	capturedMouse = on;
 }
 
 bool ScintillaWin::HaveMouseCapture() {
