--- openkomodo/src/codeintel/lib/codeintel2/lang_php.py
+++ openkomodo/src/codeintel/lib/codeintel2/lang_php.py
@@ -1177,7 +1177,10 @@ class PHPBuffer(UDLBuffer, XMLParsingBufferMixin):
     cpln_stop_chars = "~`@%^&*()=+{}]|\\;:'\",.<>?/ "
 
     def __init__(self, *args, **kwargs):
-        super(PHPBuffer, self).__init__(*args, **kwargs)
+        #sublime
+        #super(PHPBuffer, self).__init__(*args, **kwargs)
+        UDLBuffer.__init__(*args, **kwargs)
+        #endsublime
 
         if isinstance(self.accessor, KoDocumentAccessor):
             # Encourage the database to pre-scan dirs relevant to completion