--- SilverCity.orig/setup.py
+++ SilverCity/setup.py
@@ -47,7 +47,8 @@
         [os.path.join(scintilla_scr, file) for file in
             ["KeyMap.cxx",
              "Catalogue.cxx",
-            "UniConversion.cxx"]
+            "UniConversion.cxx",
+            'CharClassify.cxx']
         ]
     )
 src_files.extend(
@@ -78,7 +79,7 @@
                 'Lib/Src']

 data_files = ['CSS/default.css']
-libraries = []
+libraries = ['pcre']
 defines = []
 extra_compile_args = []
 extra_link_args = []
